### PR TITLE
crimson/os/seastore: cleanups to gc reclaim

### PR DESCRIPTION
**Pull request by cyx1231st**
This PR is mostly about cleanups. I was originally trying to improve `TransactionManager::get_extents_if_live()` to be common without segment-specific assumptions -- segment reclaiming assumes that no one else can modify the reclaiming extent in parallel except splitting/retiring it. This is why the current implementation expects to read out none or multiple extents from one allocation info.

The issue is that the transaction cannot reliably read out absent extents from backref allocation info, even though the info is correct at the read time. This is because (for example) we have to asynchronously query the lba tree before actually reading an absent logical extent out, and the original alloc-info can be out-dated as the procedure is not atomic.

The instinct solution is to load and add the extent to the transaction read-set first before doing asynchronous queries (also refer to https://github.com/ceph/ceph/pull/45519#discussion_r917219468). But it turns out that this requires some fundamental changes. Specifically, loading an extent from cache requires an `init_func` parameter. For a logical extent it requires a pin as the assumption is the lba tree was queried first. For an lba extent it requires meta information as it assumes that the parent nodes were queried beforehand.

I'm not really sure whether the described issue will be a problem for RBM with overwrite capabilities. For example when decide to load and evict a cold extent from RBM tier to a slower tier, would this leads to evicting an unrelated extent?

What do you think? @myoungwon @xxhdx1985126 @athanatos 

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/transaction_manager.cc [hunk 1]
```diff
@@ -565,11 +569,12 @@ TransactionManager::get_extents_if_live(
...
             auto pin_len = pin->get_length();
-            if (pin_paddr_seg_id != paddr_seg_id ||
-                pin_seg_paddr < paddr ||
-                pin_seg_paddr.add_offset(pin_len) > paddr.add_offset(len)) {
+            if (pin_paddr_seg_id != paddr_seg_id) {
               return seastar::now();
             }
+            // Only extent split can happen during the lookup
+            ceph_assert(pin_seg_paddr >= paddr &&
+                        pin_seg_paddr.add_offset(pin_len) <= paddr.add_offset(len));
```
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction_manager.cc:577]: Is this reasonable to you? @zhscn

Comment (zhscn) [reply src/crimson/os/seastore/transaction_manager.cc:577]: Yes

**Comment by myoungwon** [comment]
I think the issue can happen even in RBM environment in the future---probably in case of 1) tiering you mentioned or 2) defragmenter  which requires data movement to secure large aligned space?, because the procedure is non-atomic even though the transaction includes alloc-info and related information. But, it does not seem to be a problem for now in terms of RBM due to in-place allocation.

File: src/crimson/os/seastore/async_cleaner.cc [hunk 2]
```diff
@@ -905,23 +905,11 @@ AsyncCleaner::retrieve_backref_mappings(
...
-	  t, start_paddr, end_paddr
-	);
-      }).safe_then([&pin_list](auto&& list) {
-	pin_list = std::move(list);
-      });
-    }).safe_then([&pin_list] {
-      return seastar::make_ready_future<backref_pin_list_t>(std::move(pin_list));
-    });
+  return ecb->with_transaction_weak(
+      "backref_get_mappings",
```
Comment (myoungwon) [inline src/crimson/os/seastore/async_cleaner.cc]: Can I ask a question about why did you change the original transaction to weak transaction?

Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc]: There is a restriction from the current backref-tree implementation, that unlike lba tree, it doesn't support tree-read during tree-updates or concurrent tree-updates with parallel transactions. This is why our gc transactions are called in serial, and why I changed this READ transaction to weak for readability. Probably I need to add more comments here.

I'm not sure about the specific gap in the backref-tree to enable parallism, probably @xxhdx1985126 is able to elaborate.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc]: The reason why I prevented backref-tree nodes from concurrent accessing was to prevent future developers from merging in-cache backref entries in parallel or updating the backref tree in client io transactions. Because the backref tree's key is `paddr_t`, if we allow updating backref tree in client io transactions or parallel in-cache backref entries merging, in the scenario of single segmented device, all those transactions are probably gonna conflict on the same backref node, which could hurt the performance of seastore very badly.

But, when it comes to the multi-device scenario, I think we can loose that restriction a little bit so that cleaners of different devices can update the backref tree concurrently, because `paddrs` of different devices must exist in whole different sub-trees and, unless the number of devices exceeds the size of backref tree's internal node, those transactions can only conflict on the root of the backref tree. Am I right? @cyx1231st @myoungwon @athanatos

Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc]: > Because the backref tree's key is paddr_t, if we allow updating backref tree in client io transactions or parallel in-cache backref entries merging, in the scenario of single segmented device, all those transactions are probably gonna conflict on the same backref node, which could hurt the performance of seastore very badly.

I agree with the point. In addition, the backref tree is a mirror of the lba tree, so all the logical conflicts should have been addressed by the lba tree already. Further conflicts in the backref tree seem to be unnecessary.

> But, when it comes to the multi-device scenario, I think we can loose that restriction a little bit so that cleaners of different devices can update the backref tree concurrently ...

Is there really any restrictions here? Only the alloc-trim transaction is updating the backref-tree, and seems to me only the primary tier is responsible to do journaling and trimming. IIUC there is no real requirement to allow concurrent updates to the backref tree. @xxhdx1985126

Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc]: > Is there really any restrictions here? Only the alloc-trim transaction is updating the backref-tree, and seems to me only the primary tier is responsible to do journaling and trimming. IIUC there is no real requirement to allow concurrent updates to the backref tree. @xxhdx1985126

Ah, yes! hah

Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc]: Right, so we should not allow concurrent transactions between trim and reclaim right now.

Comment (myoungwon) [reply src/crimson/os/seastore/async_cleaner.cc]: Understood.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc]: > Right, so we should not allow concurrent transactions between trim and reclaim right now.

Um, actually, I think maybe we can allow this in the multi-device scenario. Because paddr mappings of different devices should be in different subtrees, the parent of which is the backref tree root. So, the alloc-trim transactions of the primary cleaner can only conflict with reclaim transactions of other cleaner when the alloc-trim transactions touch the backref tree root, which might be a relatively low-possibility event. But, I'm not sure whether it's really that low, might need more tests on this issue.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc]: > Right, so we should not allow concurrent transactions between trim and reclaim right now.

Yeah

**Comment by tchaikov** [comment]
jenkins test make check

File: src/crimson/os/seastore/async_cleaner.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc]: > Probably I need to add more comments here.

Comments added.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/crimson/os/seastore/transaction_manager.cc [hunk 3]
```diff
@@ -552,61 +557,53 @@ TransactionManager::get_extents_if_live_ret TransactionManager::get_extents_if_l
...
           DEBUGT("{} {}~{} {} is live as physical extent -- {}",
-                 t, type, laddr, len, addr, *ret);
+                 t, type, laddr, len, paddr, *ret);
         } else {
           DEBUGT("{} {}~{} {} is not live as physical extent",
-                 t, type, laddr, len, addr);
+                 t, type, laddr, len, paddr);
         }
 	std::list<CachedExtentRef> res;
 	res.emplace_back(std::move(ret));
```
Comment (zhscn) [inline src/crimson/os/seastore/transaction_manager.cc:609]: It's possible that ret is empty, so I think this line should be moved to if(ret) branch.

And line 581 is another case, it might also need to check empty before emplace_back.

Comment (cyx1231st) [reply src/crimson/os/seastore/transaction_manager.cc:609]: > It's possible that ret is empty, so I think this line should be moved to if(ret) branch.

Yeah.

> And line 581 is another case, it might also need to check empty before emplace_back.

Doesn't seem to me that the extent can be empty as the pin is retrieved and is valid. It should result in transaction conflict if another committed transaction has retired the extent at the same time.

Comment (zhscn) [reply src/crimson/os/seastore/transaction_manager.cc]: > Doesn't seem to me that the extent can be empty as the pin is retrieved and is valid. It should result in transaction conflict if another committed transaction has retired the extent at the same time.

oh, I get it.

Comment (cyx1231st) [reply src/crimson/os/seastore/transaction_manager.cc]: > It's possible that ret is empty, so I think this line should be moved to if(ret) branch.

Done, this is a fix rather than a clean-up, so added a commit for it.

File: src/crimson/os/seastore/async_cleaner.cc [hunk 4]
```diff
@@ -700,17 +700,6 @@ void AsyncCleaner::close_segment(segment_id_t segment)
...
-  journal_seq_t limit)
-{
-  return backref_manager.merge_cached_backrefs(
-    t,
-    limit,
-    config.rewrite_backref_bytes_per_cycle
-  );
-}
-
 double AsyncCleaner::calc_gc_benefit_cost(
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc:703]: Nothing is logically changed by the commit name "cleanup".

I think it is clearer to unsplit the straitforward logic about trimming. `gc_trim_alloc()` and `gc_trim_dirty()` shares similar procedure, so also make their logs consistent with better information for debugging.

Further, we might need to rearrange these cleaner transactions so that trim and reclaim won't be concurrent at the higer level with device tiers.

**Comment by cyx1231st** [comment]
jenkins test api

**Review by myoungwon — APPROVED** [review]

**Review by xxhdx1985126 — APPROVED** [review]
